src/server/debug.rs

```diff
@@ -444,6 +482,223 @@ impl Debugger {
     }
 }
 
+pub struct MvccChecker {
+    db: Arc<DB>,
+    lock_iter: DBIterator,
+    default_iter: DBIterator,
+    write_iter: DBIterator,
+    lock_fix_count: usize,
+    default_fix_count: usize,
+    write_fix_count: usize,
+}
+
+impl MvccChecker {
+    fn new(db: Arc<DB>, start_key: &[u8], end_key: &[u8]) -> Result<Self> {
+        let start_key = keys::data_key(start_key);
+        let end_key = keys::data_end_key(end_key);
+        let gen_iter = |cf: &str| -> Result<_> {
+            let from = start_key.clone();
+            let to = end_key.clone();
+            let readopts = IterOption::new(Some(from.clone()), Some(to), false).build_read_opts();
+            let handle = box_try!(get_cf_handle(db.as_ref(), cf));
+            let mut iter = DBIterator::new_cf(Arc::clone(&db), handle, readopts);
+            iter.seek(SeekKey::Start);
+            Ok(iter)
+        };
+
+        Ok(MvccChecker {
+            db: Arc::clone(&db),
+            write_iter: gen_iter(CF_WRITE)?,
+            lock_iter: gen_iter(CF_LOCK)?,
+            default_iter: gen_iter(CF_DEFAULT)?,
+            lock_fix_count: 0,
+            default_fix_count: 0,
+            write_fix_count: 0,
+        })
+    }
+
+    fn min_key(key: Option<Vec<u8>>, iter: &DBIterator, f: fn(&[u8]) -> &[u8]) -> Option<Vec<u8>> {
+        let iter_key = if iter.valid() {
+            Some(f(keys::origin_key(iter.key())).to_vec())
+        } else {
+            None
+        };
+        match (key, iter_key) {
+            (Some(a), Some(b)) => if a < b {
+                Some(a)
+            } else {
+                Some(b)
+            },
+            (Some(a), None) => Some(a),
+            (None, Some(b)) => Some(b),
+            (None, None) => None,
+        }
+    }
+
+    pub fn check_mvcc(&mut self, wb: &WriteBatch) -> Result<()> {
+        loop {
+            // Find min key in the 3 CFs.
+            let mut key = MvccChecker::min_key(None, &self.default_iter, truncate_ts);
+            key = MvccChecker::min_key(key, &self.lock_iter, |k| k);
+            key = MvccChecker::min_key(key, &self.write_iter, truncate_ts);
+
+            match key {
+                Some(key) => self.check_mvcc_key(wb, key.as_ref())?,
+                None => return Ok(()),
+            }
+        }
+    }
+
+    fn check_mvcc_key(&mut self, wb: &WriteBatch, key: &[u8]) -> Result<()> {
+        let (mut default, mut write, mut lock) = (None, None, None);
+        let (mut next_default, mut next_write, mut next_lock) = (true, true, true);
+        loop {
+            if next_default {
```

Comment:
Looks like you don't need this flag, you just need to call `next_default()` whenever you set `next_default = true`?